### DISCUSSION TITLE
Git Branching & Version Control Best Practice

### DISCUSSION THREAD

**Discussion by Jaardy** [swe-delivery/swe-tube#3] [General]
[Github Flow](https://docs.github.com/en/get-started/quickstart/github-flow)

This seems like a good candidate for early version control.   It's "lightweight".  I might argue that we could add a main ->dev ->feature pattern later on if we want to build in ci/cd.  For earlier projects, I can see how a main ->  feature pattern reduces complexity.

**Comment by accuser** [comment]
`main` -> `develop` -> `feature` is worth grasping early. The `main` branch should always be deployable, and whilst `develop` should be too, it is deployable in a development environment and context.

We can introduce a `release` branch later: `develop` -> `release` -> `main` (tagged).

**Reply by Jaardy** [reply]
Is there a danger of adding it too early?  For this project, it makes sense to build to our target of best practice.  I worry it might be too cumbersome to use too early if there isn't really a use-case.   
I suppose it's a question of identifying the optimal point at which to introduce it during Bootcamp; it would likely be for the first group project.

For this project, we're agreed `main` -> `develop` -> `feature` or should we be ambitious and plan for releases?

**Reply by accuser** [reply]
We don't *need* the release branch until we've got something to actually release... 😉

**Reply by nimonian** [reply]
Lots of apprentices find `main` -> `feature` difficult to understand, and I say this because I think `main` -> `dev` -> `feature` might actually help with that. With different branches having different roles it actually helps to understand what is the point of them.

Run tests/linting/quality checks on `dev` and continuous deployment on `main`?

**Comment by Jaardy-MV** [comment]
Apologies for reopening!
@accuser I was looking at your commits and reflecting on a shared strategy.  It's something we didn't get to cover in the bootcamp.
I was looking into the ["Atomic commit](https://dev.to/paulinevos/atomic-commits-will-help-you-git-legit-35i7)" strategy. 
Thoughts?

**Reply by accuser** [reply]
Until I got to the bit about `git push -f` I was finding myself agreeing with the sentiment of the strategy as proposed.

The `main` and `dev` (and, in the future, `release`) branches benefit from atomic commits, but this can be achieved through squashing and merging a PR from a `feature` (or other) branch. 

I think of `feature` branches as WIP - having fine-grained commits is useful because it tells a story, and it is easy to move through the history of the changes. The first PR represents the culmination of the work to date, and should pass all CI checks and code review before being merged. At the point of merge, squashing the commits results in an atomic commit into the upstream branch.

When submitting a PR, we should have confirmed that it can be merged. This doesn't mean that subsequent merges from other PRs doesn't change this, but it simply means that we need to track upstream changes. 

Or have I missed something?

**Reply by accuser** [reply]
I saw [this](https://suchdevblog.com/lessons/AtomicGitCommits.html) on my DEV feed, and it made me think again about the article you shared.

It think the thing that bothers me most about the article is this in the linked cheat sheet: _Stage your changes with `git add .`, then amend them to your last commit with `git commit --am`. Edit your commit message and/or description if desired, and write the file._ 

The author seems to be suggesting that barrelling happens by amending the previous commit. From what I can tell, this is breaking the atomic commit - it is no longer atomic, because other changes have been added. 

I can see the benefit of amending a non-pushed commit it staging was wrong, but not as a standard practice.

Maybe I'm misreading or misunderstanding.

I think I aim for atomic commits, although I've never called them that until today. They do require planning - i.e., I'm going to do this, which will mean changing that, so I'll change that first and commit, and then do this and commit, and so on. Which is A Good Thing&trade;.

**Comment by Jaardy-MV** [comment]
Glad it's got you thinking.  I saw it and realised I wasn't really in a position to comment one way or the other as my approach isn't really one at all, and I realised we probably need to address it with something.  I saw your commits and had seen the 'atomic commits' strategy and thought there was something to it.  That being said, I am approaching it from a "don't know, let's find out" approach.

**Comment by Jaardy-MV** [comment]
I'm continuing to look for info around this, but might be worth batting around next time we all catch up.

**Reply by nimonian** [reply]
https://www.youtube.com/watch?v=ihFoFqvZhCM <-- Primagean talks about atomic commits
